### PR TITLE
feature/memory

### PR DESCRIPTION
**Pull request by P403n1x87**
### Description of the Change

This change brings memory profiling capabilities to Austin. The new switch -m can be used to output memory usage metrics instead of timing information. The new switch -f can be used to output a full set of metrics including timings, memory allocations, and memory releases.

### Alternate Designs

Given the challenges presented by statistical profilers, every design choice was going to be riddled with some accuracy issues of sorts. It was decided to use resident memory as the most useful and reliable memory usage metrics as it represents the actual size taken by the process in physical memory.

### Regressions

I expect a very minimal performance impact due to the branching logic required in order to output the requested metrics. This should be totally compensated by branching prediction available on most common CPU architecture these days.

No software regressions expected.

### Verification Process

Test suite to be enhanced.

### CODE REVIEW

File: src/austin.c [hunk 1]
```diff
@@ -195,6 +235,11 @@ int main(int argc, char ** argv) {
     }
   }
 
+  if (output_file != stdout) {
+    fclose(output_file);
```
Comment (randomstuff) [inline src/austin.c]: This leave the process without any stdout file descriptor which is really unusual. Every Unix program expects to have something in stdout/stderr/stdin:

```sh
$ ./src/austin echo test
echo: write error: Bad file descriptor
```

It'd be better to open `/dev/null` and dup to it stdout.

File: src/py_proc.c [hunk 2]
```diff
@@ -672,7 +673,13 @@ py_proc__start(py_proc_t * self, const char * exec, char * argv[]) {
   #else                                                               /* UNIX */
   self->pid = fork();
   if (self->pid == 0) {
+    // If we are not writing to file we need to ensure the child process is
+    // not writing to stdout.
+    if (output_file == NULL)
+      close(STDOUT_FILENO);
```
Comment (randomstuff) [inline src/py_proc.c]: This leave the process without any stdout file descriptor which is really unusual. Every Unix program expects to have something in stdout/stderr/stdin:

```sh
$ ./src/austin echo test
echo: write error: Bad file descriptor
```

It'd be better to open `/dev/null` and dup to it stdout.

File: src/austin.c (hunk 1 above)
Comment (P403n1x87) [reply src/austin.c]: Fixed

File: src/py_proc.c (hunk 2 above)
Comment (P403n1x87) [reply src/py_proc.c]: Fixed

Comment (P403n1x87) [reply src/py_proc.c]: Fixed

**Comment by P403n1x87** [comment]
@randomstuff Thanks for your findings. I have addressed them and pushed a new commit.

**Comment by P403n1x87** [comment]
@randomstuff Thanks for your findings. I have addressed them and pushed a new commit.

File: src/py_proc.c (hunk 2 above)
Comment (P403n1x87) [reply src/py_proc.c]: Fixed

**Comment by randomstuff** [comment]
Looks good to me.

**Pull request merged**